File: CiaiControllerSDK/Services/CommunicationProviderRegistry.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.IO.Ports;
using System.Linq;
using System.Text;
using CiaiControllerSDK.Communication;
using CiaiControllerSDK.Core;
using CiaiControllerSDK.Interfaces;

namespace CiaiControllerSDK.Services
{
    /// <summary>内置连接和厂商连接的线程安全注册表。</summary>
    public static class CommunicationProviderRegistry
    {
        private static readonly ConcurrentDictionary<string, ICommunicationProvider> Providers =
            new(StringComparer.OrdinalIgnoreCase);

        static CommunicationProviderRegistry()
        {
            Register(new DelegateProvider(new[] { "tcp" }, ValidateTcp, c =>
                new TcpCommunication(c.Host, c.Port, c.ConnectTimeoutMs, c.ReadTimeoutMs, c.WriteTimeoutMs)));
            Register(new DelegateProvider(new[] { "http", "https" }, ValidateHttp, c =>
                new HttpCommunication(c.BaseUrl, c.ConnectTimeoutMs, c.Headers)));
            Register(new DelegateProvider(new[] { "serial" }, ValidateSerial, CreateSerial));
            Register(new DelegateProvider(new[] { "process", "legacy-process", "dll-process" },
                ValidateProcess, c => new ProcessCommunication(c)));
        }

        public static void Register(ICommunicationProvider provider, bool replace = false)
        {
            if (provider == null) throw new ArgumentNullException(nameof(provider));
            foreach (var type in provider.Types ?? Enumerable.Empty<string>())
            {
                if (string.IsNullOrWhiteSpace(type)) continue;
                if (replace) Providers[type.Trim()] = provider;
                else if (!Providers.TryAdd(type.Trim(), provider))
                    throw new InvalidOperationException($"通信类型已注册: {type}");
            }
        }

        public static bool IsRegistered(string type) =>
            !string.IsNullOrWhiteSpace(type) && Providers.ContainsKey(type.Trim());

        public static ICommunication Create(ConnectionConfiguration configuration)
        {
            var provider = Validate(configuration);
            return provider.Create(configuration);
        }

        public static ICommunicationProvider Validate(ConnectionConfiguration configuration)
        {
            if (configuration == null) throw new ArgumentNullException(nameof(configuration));
            if (!Providers.TryGetValue(configuration.Type?.Trim() ?? string.Empty, out var provider))
                throw new NotSupportedException(
                    $"未注册通信类型 '{configuration.Type}'。请实现ICommunicationProvider并在启动前注册。");
            ValidateCommon(configuration);
            provider.Validate(configuration);
            return provider;
        }

        private static void ValidateCommon(ConnectionConfiguration c)
        {
            if (string.IsNullOrWhiteSpace(c.Name)) throw new ArgumentException("连接名称不能为空");
            if (c.ConnectTimeoutMs <= 0 || c.ReadTimeoutMs <= 0 || c.WriteTimeoutMs <= 0)
                throw new ArgumentException($"连接 {c.Name} 的超时必须大于0");
            if (c.ResourceWaitTimeoutMs <= 0 || c.EffectiveMaxConcurrency <= 0)
                throw new ArgumentException($"连接 {c.Name} 的资源参数必须大于0");
            if (c.RetryCount < 0 || c.RetryDelayMs < 0 || c.RetryBackoff < 1)
                throw new ArgumentException($"连接 {c.Name} 的重试参数无效");
        }

        private static void ValidateTcp(ConnectionConfiguration c)
        {
            if (string.IsNullOrWhiteSpace(c.Host) || c.Port is <= 0 or > 65535)
                throw new ArgumentException($"TCP连接 {c.Name} 必须配置host和1..65535端口");
        }

        private static void ValidateHttp(ConnectionConfiguration c)
        {
            if (!Uri.TryCreate(c.BaseUrl, UriKind.Absolute, out var uri) ||
                (uri.Scheme != Uri.UriSchemeHttp && uri.Scheme != Uri.UriSchemeHttps))
                throw new ArgumentException($"HTTP连接 {c.Name} 必须配置有效baseUrl");
        }

        private static void ValidateSerial(ConnectionConfiguration c)
        {
            if (string.IsNullOrWhiteSpace(c.SerialPort))
                throw new ArgumentException($"串口连接 {c.Name} 必须配置port");
        }

        private static void ValidateProcess(ConnectionConfiguration c)
        {
            if (string.IsNullOrWhiteSpace(c.Executable))
                throw new ArgumentException($"进程连接 {c.Name} 必须配置executable");
            if (!string.IsNullOrWhiteSpace(c.WorkingDirectory) && !Directory.Exists(c.WorkingDirectory))
                throw new ArgumentException($"进程连接 {c.Name} 的workingDirectory不存在: {Path.GetFullPath(c.WorkingDirectory)}");
            var hasPath = Path.IsPathRooted(c.Executable) ||
                          c.Executable.IndexOf(Path.DirectorySeparatorChar) >= 0 ||
                          c.Executable.IndexOf(Path.AltDirectorySeparatorChar) >= 0;
            if (hasPath && !File.Exists(c.Executable))
                throw new ArgumentException($"进程连接 {c.Name} 的executable不存在: {Path.GetFullPath(c.Executable)}");
            var architecture = (c.Architecture ?? "auto").Trim().ToLowerInvariant();
            if (architecture is not ("auto" or "x86" or "x64" or "arm64"))
                throw new ArgumentException($"进程连接 {c.Name} 的architecture无效: {c.Architecture}");
            var apartment = (c.ApartmentState ?? "MTA").Trim().ToUpperInvariant();
            if (apartment is not ("STA" or "MTA"))
                throw new ArgumentException($"进程连接 {c.Name} 的apartmentState无效: {c.ApartmentState}");
            if (c.ShutdownTimeoutMs <= 0)
                throw new ArgumentException($"进程连接 {c.Name} 的shutdownTimeoutMs必须大于0");
        }

        private static ICommunication CreateSerial(ConnectionConfiguration c) =>
            new SerialCommunication(c.SerialPort, c.BaudRate, ParseParity(c.Parity), c.DataBits,
                ParseStopBits(c.StopBits), c.ReadTimeoutMs, c.WriteTimeoutMs, ParseEncoding(c.Encoding),
                ParseHandshake(c.FlowControl), c.DtrEnable, c.RtsEnable, c.DiscardInputBeforeWrite);

        private static Parity ParseParity(string value) => (value ?? "none").Trim().ToLowerInvariant() switch
        {
            "none" => Parity.None, "odd" => Parity.Odd, "even" => Parity.Even,
            "mark" => Parity.Mark, "space" => Parity.Space,
            _ => throw new ArgumentException($"不支持的串口校验方式: {value}")
        };

        private static StopBits ParseStopBits(double value)
        {
            if (Math.Abs(value - 1) < .001) return StopBits.One;
            if (Math.Abs(value - 1.5) < .001) return StopBits.OnePointFive;
            if (Math.Abs(value - 2) < .001) return StopBits.Two;
            throw new ArgumentException($"不支持的停止位: {value}");
        }

        private static Handshake ParseHandshake(string value) =>
            (value ?? "none").Trim().ToLowerInvariant() switch
            {
                "none" => Handshake.None,
                "xonxoff" or "xon/xoff" or "software" => Handshake.XOnXOff,
                "rtscts" or "rts/cts" or "hardware" => Handshake.RequestToSend,
                "rtscts+xonxoff" or "both" => Handshake.RequestToSendXOnXOff,
                _ => throw new ArgumentException($"不支持的流控方式: {value}")
            };

        private static Encoding ParseEncoding(string value) => (value ?? "utf-8").Trim().ToLowerInvariant() switch
        {
            "utf-8" or "utf8" => Encoding.UTF8, "ascii" => Encoding.ASCII,
            "unicode" or "utf-16" or "utf16" => Encoding.Unicode,
            "latin1" or "iso-8859-1" => Encoding.Latin1,
            _ => throw new ArgumentException($"不支持的串口编码: {value}")
        };

        private sealed class DelegateProvider : ICommunicationProvider
        {
            private readonly Action<ConnectionConfiguration> _validate;
            private readonly Func<ConnectionConfiguration, ICommunication> _create;
            public IEnumerable<string> Types { get; }
            public DelegateProvider(IEnumerable<string> types, Action<ConnectionConfiguration> validate,
                Func<ConnectionConfiguration, ICommunication> create)
            { Types = types; _validate = validate; _create = create; }
            public void Validate(ConnectionConfiguration configuration) => _validate(configuration);
            public ICommunication Create(ConnectionConfiguration configuration) => _create(configuration);
        }
    }
}

File: CiaiControllerSDKForJava/src/main/java/com/ciai/controller/sdk/service/CommunicationProviderRegistry.java
package com.ciai.controller.sdk.service;

import com.ciai.controller.sdk.communication.HttpCommunication;
import com.ciai.controller.sdk.communication.ProcessCommunication;
import com.ciai.controller.sdk.communication.SerialCommunication;
import com.ciai.controller.sdk.communication.TcpCommunication;
import com.ciai.controller.sdk.core.ConnectionConfiguration;
import com.ciai.controller.sdk.interface_.ICommunication;
import com.ciai.controller.sdk.interface_.ICommunicationProvider;
import com.fazecast.jSerialComm.SerialPort;

import java.net.URI;
import java.io.File;
import java.nio.charset.Charset;
import java.util.Arrays;
import java.util.Collection;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/** 内置连接和厂商连接的线程安全注册表。 */
public final class CommunicationProviderRegistry {
    private static final Map<String, ICommunicationProvider> PROVIDERS = new ConcurrentHashMap<>();
    static {
        register(provider(Arrays.asList("tcp"), new Validator(){ public void validate(ConnectionConfiguration c){
            if (blank(c.getHost()) || c.getPort() <= 0 || c.getPort() > 65535)
                throw new IllegalArgumentException("TCP connection requires host and port: " + c.getName());
        }}, new Creator(){ public ICommunication create(ConnectionConfiguration c){
            return new TcpCommunication(c.getHost(), c.getPort(), c.getConnectTimeoutMs(),
                    c.getReadTimeoutMs(), c.getWriteTimeoutMs()); }}), false);
        register(provider(Arrays.asList("http", "https"), new Validator(){ public void validate(ConnectionConfiguration c){
            try {
                URI u = URI.create(c.getBaseUrl());
                String scheme = u.getScheme();
                if (scheme == null || !(scheme.equalsIgnoreCase("http") || scheme.equalsIgnoreCase("https")))
                    throw new Exception();
            }
            catch (Exception e) { throw new IllegalArgumentException("Invalid HTTP baseUrl: " + c.getName()); }
        }}, new Creator(){ public ICommunication create(ConnectionConfiguration c){
            return new HttpCommunication(c.getBaseUrl(), c.getConnectTimeoutMs(), c.getHeaders()); }}), false);
        register(provider(Arrays.asList("serial"), new Validator(){ public void validate(ConnectionConfiguration c){
            if (blank(c.getSerialPort())) throw new IllegalArgumentException("Serial port is required: " + c.getName());
            if (c.getBaudRate() <= 0) throw new IllegalArgumentException("Serial baudRate must be positive: " + c.getName());
            if (c.getDataBits() < 5 || c.getDataBits() > 8)
                throw new IllegalArgumentException("Serial dataBits must be between 5 and 8: " + c.getName());
            stopBits(c.getStopBits());
            parity(c.getParity());
            Charset.forName(c.getEncoding());
            flowControl(c.getFlowControl());
        }}, new Creator(){ public ICommunication create(ConnectionConfiguration c){
            return new SerialCommunication(c.getSerialPort(), c.getBaudRate(), c.getDataBits(),
                    stopBits(c.getStopBits()), parity(c.getParity()), c.getReadTimeoutMs(), c.getWriteTimeoutMs(),
                    Charset.forName(c.getEncoding()), flowControl(c.getFlowControl()), c.isDtrEnable(),
                    c.isRtsEnable(), c.isDiscardInputBeforeWrite()); }}), false);
        register(provider(Arrays.asList("process", "legacy-process", "dll-process"),
                new Validator(){ public void validate(ConnectionConfiguration c){
                    if (blank(c.getExecutable())) throw new IllegalArgumentException("Process executable is required: " + c.getName());
                    if (!blank(c.getWorkingDirectory()) && !new File(c.getWorkingDirectory()).isDirectory())
                        throw new IllegalArgumentException("Process workingDirectory does not exist: " + c.getWorkingDirectory());
                    if (hasPath(c.getExecutable()) && !new File(c.getExecutable()).isFile())
                        throw new IllegalArgumentException("Process executable does not exist: " + c.getExecutable());
                    String architecture = blank(c.getArchitecture()) ? "auto" : c.getArchitecture().trim().toLowerCase(Locale.ROOT);
                    if (!(architecture.equals("auto") || architecture.equals("x86") || architecture.equals("x64") || architecture.equals("arm64")))
                        throw new IllegalArgumentException("Invalid process architecture: " + c.getArchitecture());
                    String apartment = blank(c.getApartmentState()) ? "MTA" : c.getApartmentState().trim().toUpperCase(Locale.ROOT);
                    if (!(apartment.equals("STA") || apartment.equals("MTA")))
                        throw new IllegalArgumentException("Invalid process apartmentState: " + c.getApartmentState());
                    if (c.getShutdownTimeoutMs() <= 0)
                        throw new IllegalArgumentException("Process shutdownTimeoutMs must be positive: " + c.getName());
                }}, new Creator(){ public ICommunication create(ConnectionConfiguration c){ return new ProcessCommunication(c); }}), false);
    }
    private CommunicationProviderRegistry() {}

    public static void register(ICommunicationProvider provider) { register(provider, false); }
    public static void register(ICommunicationProvider provider, boolean replace) {
        if (provider == null) throw new IllegalArgumentException("Provider is required");
        for (String type : provider.getTypes()) {
            String key = type.trim().toLowerCase(Locale.ROOT);
            if (replace) PROVIDERS.put(key, provider);
            else if (PROVIDERS.putIfAbsent(key, provider) != null)
                throw new IllegalStateException("Communication type already registered: " + type);
        }
    }
    public static boolean isRegistered(String type) { return type != null && PROVIDERS.containsKey(type.trim().toLowerCase(Locale.ROOT)); }
    public static ICommunication create(ConnectionConfiguration c) {
        ICommunicationProvider provider = validate(c);
        return provider.create(c);
    }
    public static ICommunicationProvider validate(ConnectionConfiguration c) {
        if (c == null || blank(c.getType())) throw new IllegalArgumentException("Connection type is required");
        ICommunicationProvider provider = PROVIDERS.get(c.getType().trim().toLowerCase(Locale.ROOT));
        if (provider == null) throw new IllegalArgumentException("Unregistered communication type: " + c.getType());
        validateCommon(c); provider.validate(c); return provider;
    }
    private static void validateCommon(ConnectionConfiguration c) {
        if (blank(c.getName())) throw new IllegalArgumentException("Connection name is required");
        if (c.getConnectTimeoutMs() <= 0 || c.getReadTimeoutMs() <= 0 || c.getWriteTimeoutMs() <= 0
                || c.getResourceWaitTimeoutMs() <= 0) throw new IllegalArgumentException("Invalid timeout: " + c.getName());
        if (c.getMaxConcurrency() <= 0) throw new IllegalArgumentException("maxConcurrency must be positive: " + c.getName());
        if (c.getRetryCount() < 0 || c.getRetryDelayMs() < 0 || c.getRetryBackoff() < 1)
            throw new IllegalArgumentException("Invalid retry policy: " + c.getName());
    }
    private interface Validator { void validate(ConnectionConfiguration c); }
    private interface Creator { ICommunication create(ConnectionConfiguration c); }
    private static ICommunicationProvider provider(final Collection<String> types, final Validator v, final Creator cr) {
        return new ICommunicationProvider() {
            public Collection<String> getTypes(){return types;}
            public void validate(ConnectionConfiguration c){v.validate(c);}
            public ICommunication create(ConnectionConfiguration c){return cr.create(c);}
        };
    }
    private static boolean blank(String v){return v==null||v.trim().isEmpty();}
    private static boolean hasPath(String v){return v.indexOf('/')>=0||v.indexOf('\\')>=0||new File(v).isAbsolute();}
    private static int parity(String v){String x=v==null?"none":v.toLowerCase(Locale.ROOT); if(x.equals("odd"))return SerialPort.ODD_PARITY;if(x.equals("even"))return SerialPort.EVEN_PARITY;if(x.equals("mark"))return SerialPort.MARK_PARITY;if(x.equals("space"))return SerialPort.SPACE_PARITY;if(x.equals("none"))return SerialPort.NO_PARITY;throw new IllegalArgumentException("Invalid parity: "+v);}
    private static int stopBits(double v){if(Math.abs(v-1)<.001)return SerialPort.ONE_STOP_BIT;if(Math.abs(v-1.5)<.001)return SerialPort.ONE_POINT_FIVE_STOP_BITS;if(Math.abs(v-2)<.001)return SerialPort.TWO_STOP_BITS;throw new IllegalArgumentException("Invalid stop bits: "+v);}
    private static int flowControl(String v){String x=v==null?"none":v.toLowerCase(Locale.ROOT);if(x.equals("none"))return SerialPort.FLOW_CONTROL_DISABLED;if(x.equals("xonxoff")||x.equals("xon/xoff")||x.equals("software"))return SerialPort.FLOW_CONTROL_XONXOFF_IN_ENABLED|SerialPort.FLOW_CONTROL_XONXOFF_OUT_ENABLED;if(x.equals("rtscts")||x.equals("rts/cts")||x.equals("hardware"))return SerialPort.FLOW_CONTROL_RTS_ENABLED|SerialPort.FLOW_CONTROL_CTS_ENABLED;if(x.equals("both")||x.equals("rtscts+xonxoff"))return SerialPort.FLOW_CONTROL_RTS_ENABLED|SerialPort.FLOW_CONTROL_CTS_ENABLED|SerialPort.FLOW_CONTROL_XONXOFF_IN_ENABLED|SerialPort.FLOW_CONTROL_XONXOFF_OUT_ENABLED;throw new IllegalArgumentException("Invalid flow control: "+v);}
}
